datafusion/proto/src/physical_plan/to_proto.rs

```diff
@@ -374,228 +378,234 @@ fn aggr_expr_to_aggr_fn(expr: &dyn AggregateExpr) -> Result<AggrFn> {
+            let fun: protobuf::ScalarFunction = (&fun).try_into()?;
+
             Ok(protobuf::PhysicalExprNode {
-                expr_type: Some(protobuf::physical_expr_node::ExprType::Cast(Box::new(
-                    protobuf::PhysicalCastNode {
-                        expr: Some(Box::new(cast.expr().clone().try_into()?)),
-                        arrow_type: Some(cast.cast_type().try_into()?),
+                expr_type: Some(protobuf::physical_expr_node::ExprType::ScalarFunction(
+                    protobuf::PhysicalScalarFunctionNode {
+                        name: expr.name().to_string(),
+                        fun: fun.into(),
+                        args,
+                        return_type: Some(expr.return_type().try_into()?),
                     },
-                ))),
-            })
-        } else if let Some(cast) = expr.downcast_ref::<TryCastExpr>() {
-            Ok(protobuf::PhysicalExprNode {
-                expr_type: Some(protobuf::physical_expr_node::ExprType::TryCast(
-                    Box::new(protobuf::PhysicalTryCastNode {
-                        expr: Some(Box::new(cast.expr().clone().try_into()?)),
-                        arrow_type: Some(cast.cast_type().try_into()?),
-                    }),
                 )),
             })
-        } else if let Some(expr) = expr.downcast_ref::<ScalarFunctionExpr>() {
-            let args: Vec<protobuf::PhysicalExprNode> = expr
-                .args()
-                .iter()
-                .map(|e| e.to_owned().try_into())
-                .collect::<Result<Vec<_>, _>>()?;
-            if let Ok(fun) = BuiltinScalarFunction::from_str(expr.name()) {
-                let fun: protobuf::ScalarFunction = (&fun).try_into()?;
-
-                Ok(protobuf::PhysicalExprNode {
-                    expr_type: Some(
-                        protobuf::physical_expr_node::ExprType::ScalarFunction(
-                            protobuf::PhysicalScalarFunctionNode {
-                                name: expr.name().to_string(),
-                                fun: fun.into(),
-                                args,
-                                return_type: Some(expr.return_type().try_into()?),
-                            },
-                        ),
-                    ),
-                })
-            } else {
-                Ok(protobuf::PhysicalExprNode {
-                    expr_type: Some(protobuf::physical_expr_node::ExprType::ScalarUdf(
-                        protobuf::PhysicalScalarUdfNode {
-                            name: expr.name().to_string(),
-                            args,
-                            return_type: Some(expr.return_type().try_into()?),
-                        },
-                    )),
-                })
-            }
-        } else if let Some(expr) = expr.downcast_ref::<LikeExpr>() {
+        } else {
+            let mut buf = Vec::new();
+            // let _ = codec.try_encode_udf(, &mut buf);
```

Comment:
I have a question, how can I get a `ScalarUDF` from a `ScalarFunctionExpr`?
Given its current structure, I would think change the fields in `ScalarFunctionExpr` may be a method.

Reply:
I also tried using `create_udf`, but it seems can not work.

Reply:
I'm assuming this is what you mean in your below comment but I think we need to change 
```
pub struct ScalarFunctionExpr {
    fun: ScalarFunctionImplementation,
```

to 

```
pub struct ScalarFunctionExpr {
    fun: ScalarFunctionDefinition,
```

@alamb This would technically be a breaking change since `ScalarFunctionExpr` has a public method
```
pub fn fun(&self) -> &ScalarFunctionImplementation {
        &self.fun
}
```

We could technically do this is in a backwards compatible way I think by creating a corresponding `ScalarFunctionImplementation` from the `ScalarFunctionImplementation` or we could just change the public API. Seems to me like this would probably not  be a very disruptive breaking change  since `ScalarFunctionImplementation` is just an opaque function pointer so unlikely this is being used in application code in a way that wouldn't be easily adaptable if it started returning `&ScalarFunctionDefinition` instead

Reply:
Thanks for your review! @thinkharderdev 
Yeah, the change is what I thought. And I also think some other fields in `ScalarFunctionExpr` can be removed like `name` and `return_type`.

Yeah, I think create a ScalarFunctionImplementation from ScalarFunctionDefinition is a method. Given the `BuiltinScalarFunction` will be removed and we can get the `inner` of a ScalarUDF which is a `ScalarUDFImpl`, and we can do something like. 
```
let captured = inner.clone();
Arc::new(move |args| captured.invoke(args))
```
I'd like to see the opinions from you and @alamb :)